constraints/cpu.md

```diff
@@ -157,6 +157,83 @@ docker run --cpus 4 -ti debian bash -c "nproc; cat /sys/fs/cgroup/cpu,cpuacct/cp
+only the workload is placed, the second layer is in the host that is more complex and
+might contain more than one process and task (thread) depending of the number of
+containers per POD and vCPUs per container. Following diagram represents a nginx container
+created with docker with the default number of vcpus.
+
+
+```
+docker run -dt --runtime=kata-runtime nginx
+
+
+       .-------.
+       | nginx |
+    .--'-------'---.  .------------.
+    | Guest Cgroup |  | Kata agent |
+  .-'--------------'--'------------'.    .-----------.
+  |  Thread: Hypervisor's vCPU 0    |    | Kata Shim |
+ .'---------------------------------'.  .'-----------'.
+ |             Tasks                 |  |  Processes  |
+.'-----------------------------------'--'-------------'.
+|                    Host Cgroup                       |
+'------------------------------------------------------'
+```
+
+In next sections is explained what is the difference between processes and tasks and why
+only hypervisor's vCPUs are constrained.
+
+### cpu cgroups in the guess
+
+Only the workload process including all its threads are placed into cpu cgroups, this means
+that kata agent and systemd run without constraints in the guest.
+
+#### CPU pinning (cpuset cgroup)
+
+Kata Containers try to apply and honor the cgroups but sometimes that's not possible, this
+is the case of cpuset cgroup where the number of virtual CPUs (guest) does not match with
+the actual number of physical CPUs (host).
+In Kata Containers to have a good performance and small memory footprint, the resources are
+hot added when they are needed, therefore the number of virtual resources is not the same
+than the number of physical resources. The problem with this approach is that it's not possible
+to pin a process on a specific resource that is not present in the guest. To deal with this
+limitation and do not fail when the container is being created, Kata Containers do not apply
+the constraint in the first layer (guest) if the resource does not exist in the guest, but it
+is applied in the second layer (host) where the hypervisor is running. In next sections is
+explained with more details what parts of the hypervisor are constrained.
+
+### cpu cgroups in the host
+
+In Kata Containers the workloads run in a virtual machine that is managed and represented by a
+hypervisor running in the host, like other processes the hypervisor might use threads to realize
+several tasks, for example IO and Network operations, one of the most important uses for the
+threads is as vCPUs, the processes running in the guest see these vCPUs as physical CPUs, while
+in the host those vCPU are just threads that are part of a process, this is the key to ensure
+workloads consumes only the amount of CPU resources that were assigned to it without impacting
+other operations. From user perspective the easier approach to implement it would be to take the
```

Comment:
Wow, the sentence ends here :) I'd definitely break this up into a number of smaller, more easily digestible sentences.